android_jni/avifandroidjni/src/main/jni/libavif_jni.cc

```diff
@@ -0,0 +1,150 @@
+  JNIEXPORT RETURN_TYPE Java_com_avif_android_AvifDecoder_##NAME( \
+      JNIEnv* env, jobject /*thiz*/, ##__VA_ARGS__);              \
+  }                                                               \
+  JNIEXPORT RETURN_TYPE Java_com_avif_android_AvifDecoder_##NAME( \
+      JNIEnv* env, jobject /*thiz*/, ##__VA_ARGS__)
+
+namespace {
+
+jfieldID global_info_width;
+jfieldID global_info_height;
+
+// RAII wrapper function that properly frees the decoder related objects on
+// destruction.
+struct AvifDecoderWrapper {
+ public:
+  AvifDecoderWrapper() = default;
+  // Not copyable or movable.
+  AvifDecoderWrapper(const AvifDecoderWrapper&) = delete;
+  AvifDecoderWrapper& operator=(const AvifDecoderWrapper&) = delete;
+
+  ~AvifDecoderWrapper() {
+    if (image != nullptr) {
+      avifImageDestroy(image);
+    }
+    if (decoder != nullptr) {
+      avifDecoderDestroy(decoder);
+    }
+  }
+
+  avifDecoder* decoder = nullptr;
+  avifImage* image = nullptr;
+};
+
+}  // namespace
+
+jint JNI_OnLoad(JavaVM* vm, void* /*reserved*/) {
+  JNIEnv* env;
+  if (vm->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION_1_6) != JNI_OK) {
+    return -1;
+  }
+  const jclass info_class = env->FindClass("com/avif/android/AvifDecoder$Info");
+  global_info_width = env->GetFieldID(info_class, "width", "I");
+  global_info_height = env->GetFieldID(info_class, "height", "I");
+  return JNI_VERSION_1_6;
+}
+
+FUNC(jboolean, isAvifStream, jobject encoded, int length) {
+  const uint8_t* const buffer =
+      reinterpret_cast<const uint8_t*>(env->GetDirectBufferAddress(encoded));
+  const avifROData avif = {buffer, static_cast<size_t>(length)};
+  return avifPeekCompatibleFileType(&avif);
+}
+
+FUNC(jboolean, getInfo, jobject encoded, int length, jobject info) {
+  const uint8_t* const buffer =
+      reinterpret_cast<const uint8_t*>(env->GetDirectBufferAddress(encoded));
+  AvifDecoderWrapper decoder;
+  decoder.decoder = avifDecoderCreate();
+  if (decoder.decoder == nullptr) {
+    LOGE("Failed to create AVIF Decoder.");
+    return false;
+  }
+  avifResult res = avifDecoderSetIOMemory(decoder.decoder, buffer, length);
+  if (res != AVIF_RESULT_OK) {
+    LOGE("Failed to set AVIF memory");
+    return false;
+  }
+  res = avifDecoderParse(decoder.decoder);
+  if (res != AVIF_RESULT_OK) {
+    LOGE("Failed to parse AVIF image: %s.", avifResultToString(res));
+    return false;
+  }
+  env->SetIntField(info, global_info_width, decoder.decoder->image->width);
+  env->SetIntField(info, global_info_height, decoder.decoder->image->height);
+  return true;
+}
+
+FUNC(jboolean, decode, jobject encoded, int length, jobject bitmap) {
+  const uint8_t* const buffer =
+      reinterpret_cast<const uint8_t*>(env->GetDirectBufferAddress(encoded));
+  AvifDecoderWrapper decoder;
+  decoder.image = avifImageCreateEmpty();
+  if (decoder.image == nullptr) {
+    LOGE("Failed to create AVIF image.");
+    return false;
+  }
+  decoder.decoder = avifDecoderCreate();
```

Comment:
Note that Chrome sets the following AVIFDecoder fields:

```
    decoder_->maxThreads = 2;

    // Chrome doesn't use XMP and Exif metadata. Ignoring XMP and Exif will
    // ensure avifDecoderParse() isn't waiting for some tiny Exif payload hiding
    // at the end of a file.
    decoder_->ignoreXMP = AVIF_TRUE;
    decoder_->ignoreExif = AVIF_TRUE;

    // Turn off libavif's 'clap' (clean aperture) property validation. (We
    // ignore the 'clap' property.)
    decoder_->strictFlags &= ~AVIF_STRICT_CLAP_VALID;
    // Allow the PixelInformationProperty ('pixi') to be missing in AV1 image
    // items. libheif v1.11.0 or older does not add the 'pixi' item property to
    // AV1 image items. (This issue has been corrected in libheif v1.12.0.) See
    // crbug.com/1198455.
    decoder_->strictFlags &= ~AVIF_STRICT_PIXI_REQUIRED;
```

Since the Java `getInfo()` and `decode()` methods do not return the XMP and Exif metadata right now, we can set `ignoreXMP` and `ignoreExif` to `AVIF_TRUE` in both `getInfo()` and `decode()`.

We can look into how to expose `strictFlags` in Java later.

I am not sure whether we should expose `maxThreads` in Java. We can consider setting `maxThreads` to 2 in the `decode()` method.

Reply:
I have set ignoreXMP and ignoreExif.

We can expose more functions in the API to allow threads in the future. I will leave that out of this PR.